File: erpnext/accounts/report/e_payment_report/e_payment_report.py
# Copyright (c) 2023, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
	if filters.payment_type == "Bank Payment":
		columns, data = get_column(), get_data(filters)
	else:
		columns, data = get_utility_column(), get_utility_data(filters)
	return columns, data

def get_column():
	columns = [
		("Transaction") + ":Link/Bank Payment:100",
		("Transaction Type") + ":Data:150",
		("Transaction ID") + ":Data:150",
		("Transaction Date") + ":Date:100",
		("Transaction Reference") + ":Data:100",
		("Supplier") + ":Link/Supplier:120",
		("Beneficiary Name") + ":Data:150",
		("Beneficiary Bank Acc No.") + ":Data:150",
		("Total Amount") + ":Currency:100",
		("Status") + ":Data:100",
		("Journal No.") + ":Data:100",
		("PI Number") + ":Data:100"
	]
	return columns

def get_data(filters):
	cond = get_condition(filters)
	data = frappe.db.sql("""
		SELECT 
			bp.name, bpi.transaction_type, bpi.transaction_id, bpi.transaction_date, bpi.transaction_reference, bpi.supplier, bpi.beneficiary_name, bpi.bank_account_no, bpi.amount, bpi.status, bpi.bank_journal_no, bpi.pi_number
		FROM `tabBank Payment` bp, `tabBank Payment Item` bpi
		WHERE bp.name=bpi.parent
		{condition}
	""".format(condition=cond))
	return data

def get_utility_column():
	columns = [
		("Utility Service Type") + ":Data:150",
		("Party") + ":Data:150",
		("Branch Expense Account") + ":Data:200",
		("Paid From Bank Account") + ":Data:200",
		("Debit Account") + ":Link/Account:200",
		("Unique Identity Code") + ":Data:100",
		("Outstanding Amount") + ":Data:100",
		("Payment Status") + ":Data:100",
		("Payment API Response") + ":Data:400",
		("Create Direct Payment") + ":Check:100",
		("TDS Applicable") + ":Check:100",
		("PI Number")
	]
	return columns

def get_utility_data(filters):
	cond = get_condition(filters)
	data = frappe.db.sql("""
		SELECT 
			uti.utility_service_type, uti.party, ut.expense_account, ut.bank_account, uti.debit_account, uti.consumer_code, uti.outstanding_amount, uti.payment_status, uti.payment_response_msg, uti.create_direct_payment, uti.tds_applicable, uti.pi_number
		FROM `tabUtility Bill` ut, `tabUtility Bill Item` uti 
		WHERE ut.name=uti.parent
		{condition}
	""".format(condition=cond))
	return data

def get_condition(filters):
	conds = ""
	if filters.transaction_type:
		conds += "and bpi.transaction_type='{}'".format(filters.transaction_type)
	if filters.supplier:
		conds += "and bpi.supplier='{}'".format(filters.supplier)
	if filters.status:
		conds += "and bpi.status='{}'".format(filters.status)
	if filters.party:
		conds = "and uti.party='{}'".format(filters.party)
	if filters.branch:
		if filters.payment_type == "Bank Payment":
			conds = "and bp.branch = '{}'".format(filters.branch)
		else:
			conds = "and ut.branch = '{}'".format(filters.branch)
	return conds

File: erpnext/accounts/report/e_payment_report/e_payment_report.js
// Copyright (c) 2023, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["e-Payment Report"] = {
	"onload": function (query_report) {
    	query_report.filters_by_name.party.toggle(false);
    },
	"filters": [
		{
            fieldname: "payment_type",
            label: "Payment Type",
            fieldtype: "Select",
            options: ["Bank Payment", "Utility Bill Payment"],
            default: "Bank Payment",
            on_change: function(query_report){
                var payment_type = query_report.get_values().payment_type;
                var transaction_type = query_report.filters_by_name["transaction_type"];
                var supplier = query_report.filters_by_name["supplier"];
                var status = query_report.filters_by_name["status"];
                var party = query_report.filters_by_name['party'];

                if (payment_type == 'Utility Bill Payment'){
                    transaction_type.toggle(false);
                    supplier.toggle(false);
                    status.toggle(false);
                    party.toggle(true);
                }
                else{
                    transaction_type.toggle(true);
                    supplier.toggle(true);
                    status.toggle(true);
                    party.toggle(false);
                }
                query_report.refresh()
                transaction_type.refresh()
                supplier.refresh()
                status.refresh()
                party.refresh()
            }
        },
        {
            fieldname: "transaction_type",
            label: "Transaction Type",
            fieldtype: "Select",
            options: ["","Direct Payment", "Journal Entry", "Payment Entry", "Transporter Payment", "Overtime Application", "EME Payment", "Salary Slip", "Employee Loan Payment", "LTC", "Bonus", "PBVA"]
        },
        {
            fieldname: "supplier",
            label: "Supplier",
            fieldtype: "Link",
            options: "Supplier"
        },
        {
            fieldname: "status",
            label: "Status",
            fieldtype: "Select",
            options:["","Completed", "Pending", "Draft", "Waiting for Verification", "Waiting Approval", "Approved", "Rejected", "Failed", "Partial Payment", "Cancelled", "In progress", "Upload Failed", "Waiting Acknowledgement", "Processing Acknowledgement"]
        },
        {
            fieldname: "party",
            label: "Party",
            fieldtype: "Link",
            options: "Supplier"
        },
        {
            fieldname: "branch",
            label: "Branch",
            fieldtype: "Link",
            options: "Branch"
        }

	]
};
